python/samples/concepts/auto_function_calling/chat_completion_with_function_calling.py

```diff
@@ -0,0 +1,166 @@
+# - Services.ANTHROPIC
+# - Services.BEDROCK
+# - Services.GOOGLE_AI
+# - Services.MISTRAL_AI
+# - Services.OLLAMA
+# - Services.ONNX
+# - Services.VERTEX_AI
+# Please make sure you have configured your environment correctly for the selected chat completion service.
+chat_completion_service, request_settings = get_chat_completion_service_and_request_settings(Services.OPENAI)
+
+# Configure the function choice behavior. Here, we set it to Auto with auto_invoke=True.
+# - If `auto_invoke=True`, the model will automatically choose and call functions as needed.
+# - If `auto_invoke=False`, the model may return tool call instructions that you must handle and call manually.
+request_settings.function_choice_behavior = FunctionChoiceBehavior.Auto(auto_invoke=True)
+
+kernel.add_service(chat_completion_service)
+
+# Pass the request settings to the kernel arguments.
+arguments = KernelArguments(settings=request_settings)
+
+# Create a chat history to store the system message, initial messages, and the conversation.
+history = ChatHistory()
+history.add_system_message(system_message)
+history.add_user_message("Hi there, who are you?")
+history.add_assistant_message("I am Mosscap, a chat bot. I'm trying to figure out what people need.")
+
+
+def print_tool_calls(message: ChatMessageContent) -> None:
+    """
+    A helper function to pretty print the tool calls found in a ChatMessageContent message.
+    This is useful when auto tool invocation is disabled and the model returns calls that you must handle.
+    """
+    items = message.items
+    formatted_tool_calls = []
+    for i, item in enumerate(items, start=1):
+        if isinstance(item, FunctionCallContent):
+            tool_call_id = item.id
+            function_name = item.name
+            function_arguments = item.arguments
+            formatted_str = (
+                f"tool_call {i} id: {tool_call_id}\n"
+                f"tool_call {i} function name: {function_name}\n"
+                f"tool_call {i} arguments: {function_arguments}"
+            )
+            formatted_tool_calls.append(formatted_str)
+    if len(formatted_tool_calls) > 0:
+        print("\n[Tool calls returned by the model]:\n" + "\n\n".join(formatted_tool_calls))
+    else:
+        print("\n[No tool calls returned by the model]")
+
+
+async def chat() -> bool:
+    """
+    Continuously prompt the user for input and show the assistant's response.
+    Type 'exit' to exit.
+    """
+    try:
+        user_input = input("User:> ")
+    except (KeyboardInterrupt, EOFError):
+        print("\n\nExiting chat...")
+        return False
+
+    if user_input.lower().strip() == "exit":
+        print("\n\nExiting chat...")
+        return False
+
+    arguments["user_input"] = user_input
+    arguments["chat_history"] = history
+
+    # Handle non-streaming responses
+    result = await kernel.invoke(chat_function, arguments=arguments)
+
+    # If function calls are returned and auto invoking is off, we must show them.
```

Comment:
To emphasize the benefits of auto function invocation, we can break down the sample even further:
1. Manually invocation (We probably don't need to show how to manually invoke the functions. Many other SDKs don't show that on their documentations neither).
2. Auto invocation.